orm/serial_bucket.go

```diff
@@ -0,0 +1,358 @@
+	// get index
+	info := smb.getIndexInfo(indexName)
+	if info == nil {
+		return nil, errors.Wrapf(errors.ErrDatabase, "no index with name %s", indexName)
+	}
+
+	dbPrefix := append(info.prefix, prefix...)
+	start, end := prefixRange(dbPrefix)
+
+	var rawIter weave.Iterator
+	var err error
+	if reverse {
+		rawIter, err = db.ReverseIterator(start, end)
+		if err != nil {
+			return nil, errors.Wrap(err, "reverse prefix scan")
+		}
+	} else {
+		rawIter, err = db.Iterator(start, end)
+		if err != nil {
+			return nil, errors.Wrap(err, "prefix scan")
+		}
+	}
+
+	return &indexSerialModelIterator{
+		iterator:     rawIter,
+		bucketPrefix: smb.b.DBKey(nil),
+		unique:       info.unique,
+		kv:           db,
+	}, nil
+}
+
+func (smb *serialModelBucket) ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, destination SerialModelSlicePtr) error {
+	objs, err := smb.b.GetIndexed(db, indexName, key)
+	if err != nil {
+		return err
+	}
+	if len(objs) == 0 {
+		return nil
+	}
+
+	dest := reflect.ValueOf(destination)
+	if dest.Kind() != reflect.Ptr {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of SerialModels")
+	}
+	if dest.IsNil() {
+		return errors.Wrap(errors.ErrImmutable, "got nil pointer")
+	}
+	dest = dest.Elem()
+	if dest.Kind() != reflect.Slice {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of SerialModels")
+	}
+
+	// It is allowed to pass destination as both []MySerialModel and []*MySerialModel
+	sliceOfPointers := dest.Type().Elem().Kind() == reflect.Ptr
+
+	allowed := dest.Type().Elem()
+	if sliceOfPointers {
+		allowed = allowed.Elem()
+	}
+	if smb.model != allowed {
+		return errors.Wrapf(errors.ErrType, "this bucket operates on %s serialmodel and cannot return %s", smb.model, allowed)
+	}
+
+	for _, obj := range objs {
+		if obj == nil || obj.Value() == nil {
+			continue
+		}
+
+		val := reflect.ValueOf(obj.Value())
+		if err := val.Interface().(SerialModel).SetPrimaryKey(obj.Key()); err != nil {
+			return errors.Wrap(errors.ErrType, "cannot set ID")
+		}
+
+		if !sliceOfPointers {
+			val = val.Elem()
+		}
+		// store the key on the SerialModel
+		dest.Set(reflect.Append(dest, val))
+	}
+	return nil
+
+}
+
+func (smb *serialModelBucket) Create(db weave.KVStore, m SerialModel) error {
+	err := smb.enforceType(m)
+	if err != nil {
+		return errors.Wrap(err, "model type is not supported")
+	}
+
+	if err := m.Validate(); err != nil {
+		return errors.Wrap(err, "invalid serialmodel")
+	}
+
+	// If key is provided use it otherwise generate auto-incremented key
+	key := m.GetPrimaryKey()
+	if len(key) == 0 {
+		key, err = smb.idSeq.NextVal(db)
+		if err != nil {
+			return errors.Wrap(err, "ID sequence")
+		}
+	}
+
+	obj := NewSimpleObj(key, m)
+	if err := smb.b.Save(db, obj); err != nil {
+		return errors.Wrap(err, "cannot create in the database")
+	}
+	// after serialization, return original/generated key on SerialModel
+	if err := m.SetPrimaryKey(key); err != nil {
```

Comment:
req: you should store the key with the model before persisting the object. It will save you all the work in setting it later after load. Any modification on objects after load may get you into trouble with Proofs.

Reply:
Hmm... interesting. I guess it does waste a bit of space, but then makes the whole module much clearer.

The object is just like a model from sql / orm land - all fields.
Then we calculate the PrimaryKey as well as each index from it.

I think this adds a bit of space overhead, but for items with auto-increment primary key, not too bad. I wonder about cases when the key is already 20/32 bytes (eg. addresses). But in general, a very nice simplification.